### DISCUSSION TITLE
LHS/RHS type hint difference

### DISCUSSION THREAD

**Discussion by Molkree** [microsoft/pyright#3926] [Q&A] [answered]
I would expect that the two following snippets of code would produce the same inferred type for `my_list`:
```py
my_list = list[int]()  # list[int]
my_list = []  # list[Unknown]
```
```py
my_list: list[int] = []  # list[int]
my_list = []  # list[int]
```

What am I missing here? I don't see any examples of the first syntax in the [type inference doc](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md), and I just assumed previously it would be equal to the second syntax.

**Comment by erictraut** [comment, marked as answer]
The two cases are very different. In the second, you have declared the type of `my_list`. Its type is narrowed by assignment, and bidirectional type inference is used to determine the type of the assigned expression.

In the first example, you have not declared its type, so its local type is defined by each assignment without the benefit of any bidirectional type inference.

For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md) and [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md).
